test/IR/affine-map.mlir

```diff
@@ -174,6 +174,9 @@
 // CHECK: #map{{[0-9]+}} = (d0) -> (d0 - (d0 + 1))
 #map53 = (d0) -> (d0 - (d0 + 1))
 
+// CHECK: #map{{[0-9]+}} = (d0)[s0] -> ((-s0) floordiv 4, d0 floordiv -1)
+#map54 = (d0)[s0] -> (-s0 floordiv 4, d0 floordiv -1)
```

Comment:
The RHS of floordiv is expected to be positive according to the spec. I'm assuming that's the reason why the code originally assumed the -1 RHS implied multiplication.  Maybe we should fix the affine verifier instead to disallow negative constants with floordiv/ceildiv/mod.

Reply:
> The RHS of floordiv is expected to be positive according to the spec. I'm assuming that's the reason why the code originally assumed the -1 RHS implied multiplication. 

No, that was due to another bug. The change in lib/IR/AsmPrinter.cpp in this commit makes it clear: d0 floordiv -1 or d0 ceildiv -1 was being turned into -d0 due to a missing guard!

> Maybe we should fix the affine verifier instead to disallow negative constants with floordiv/ceildiv/mod.

Negative constants can appear later in the IR as a result of simplification and they could appear on branches that are never taken (basically dead code that is not yet marked dead). As such, they have to be just preserved: this includes floordiv/ceildiv/mod w.r.t zero as well besides negative values. Such expressions are left as is and aren't simplified or folded.